### PR TITLE
ASoC: SOF: Intel: hda: add generic hda codec support

### PR DESCRIPTION
**Pull request by kv2019i**
Fix for:
https://github.com/thesofproject/linux/issues/1807

 ASoC: SOF: Intel: hda: add generic hda codec support

Add support for using generic codec driver with SOF. Generic driver
is used if:
 - snd_sof_intel_hda_common.hda_model="generic" is set, or
 - fallback if no other codec driver if found

BugLink: #1807

To force this option to be used, put following in
/etc/modprobe.d/sof.conf:

options snd_sof_intel_hda_common hda_model="generic"

### CODE REVIEW

File: sound/soc/sof/intel/hda-codec.c [hunk 1]
```diff
@@ -108,6 +134,12 @@ static int hda_codec_probe(struct snd_sof_dev *sdev, int address,
 
 	hda_priv->codec.bus = hbus;
 	hdev = &hda_priv->codec.core;
+	codec = &hda_priv->codec;
+
+	if (is_generic_config(hbus))
+		codec->probe_id = HDA_CODEC_ID_GENERIC;
+	else
+		codec->probe_id = 0;
```
Comment (plbossart) [inline sound/soc/sof/intel/hda-codec.c]: @kv2019i did you mean to say that the only way to use the generic model is through the use of a kernel module parameter?

I am not sure what prevents us from doing the same as legacy, and use the generic part as a fallback?

Comment (kv2019i) [reply sound/soc/sof/intel/hda-codec.c]: @plbossart No, I just run out of time. I wanted to get a patch out to try by end-of-week and this was my late Friday effort for that. We should have proper fallback for generic and it's just a matter how to cleanly do that in SOF.

Comment (plbossart) [reply sound/soc/sof/intel/hda-codec.c]: ok, thanks for the precision!

**Comment by kv2019i** [comment]
Just an intermediate update now to get CI results for current state. It seems we can reuse most of the code from legacy driver and effectively remove code from SOF side.

**Comment by kv2019i** [comment]
This is ready for review now. I spent some time trying export the functionality from legacy driver (hda_bind.c). I got a version that passed all basic tests, but upon closer scrutiny, there were corner cases that would still fail. Basic problem is that in ASoC, we initialize the "hda_codec" instances later (in soc/codecs/hdac_hda.c). The code for driver binding in legacy driver on the other hand relies on  codec structure to be initialized. This seems a bit too dangerous.. we would be reusing functions but passing an object that is not in same state as it is in the legacy driver.

Long story short, increasing reuse for this area of code, require much more thinking. Given this problem is rather pressing (no audio on laptops that have a generic HDA codec), I think we should proceed and try to get this to 5.8.

**Comment by kv2019i** [comment]
@plbossart wrote:
> If you want to hit 5.8 you should hurry (merge window likely this week-end) and change the title to 'fix generic hda codec support')

Indeed. I'll merge this and send out to the list right away.
